pkg/rbd/controllerserver.go

```diff
@@ -265,93 +224,75 @@ func (cs *ControllerServer) DeleteVolume(ctx context.Context, req *csi.DeleteVol
+	}
 	volumeIDMutex.LockKey(volumeID)
-
 	defer func() {
 		if err := volumeIDMutex.UnlockKey(volumeID); err != nil {
 			klog.Warningf("failed to unlock mutex volume:%s %v", volumeID, err)
 		}
 	}()
 
 	rbdVol := &rbdVolume{}
-	if err := cs.MetadataStore.Get(volumeID, rbdVol); err != nil {
-		if err, ok := err.(*util.CacheEntryNotFound); ok {
-			klog.V(3).Infof("metadata for volume %s not found, assuming the volume to be already deleted (%v)", volumeID, err)
+	if err := genVolFromVolID(rbdVol, volumeID, req.GetSecrets()); err != nil {
+		// If image key is missing, there can be no unreserve as request name is unknown, and also
+		// means there is no image, hence return success
+		if _, ok := err.(util.ErrKeyNotFound); ok {
 			return &csi.DeleteVolumeResponse{}, nil
 		}
 
-		return nil, err
-	}
-
-	volName := rbdVol.VolName
-	// Deleting rbd image
-	klog.V(4).Infof("deleting volume %s", volName)
-	if err := deleteRBDImage(rbdVol, rbdVol.AdminID, req.GetSecrets()); err != nil {
-		// TODO: can we detect "already deleted" situations here and proceed?
-		klog.V(3).Infof("failed to delete rbd image: %s/%s with error: %v", rbdVol.Pool, volName, err)
-		return nil, status.Error(codes.Internal, err.Error())
-	}
-
-	if err := cs.MetadataStore.Delete(volumeID); err != nil {
-		return nil, status.Error(codes.Internal, err.Error())
-	}
-
-	delete(rbdVolumes, volumeID)
-	return &csi.DeleteVolumeResponse{}, nil
-}
-
-// ListVolumes returns a list of volumes stored in memory
-func (cs *ControllerServer) ListVolumes(ctx context.Context, req *csi.ListVolumesRequest) (*csi.ListVolumesResponse, error) {
-	var startToken int
-	if err := cs.Driver.ValidateControllerServiceRequest(csi.ControllerServiceCapability_RPC_LIST_VOLUMES); err != nil {
-		klog.Warningf("invalid list volume req: %v", req)
-		return nil, err
-	}
-
-	//validate starting token if present
-	if len(req.GetStartingToken()) > 0 {
-		i, parseErr := strconv.ParseUint(req.StartingToken, 10, 32)
-		if parseErr != nil {
-			return nil, status.Errorf(codes.Aborted, "invalid starting token %s", parseErr.Error())
-		}
-		//check starting Token is greater than list of rbd volumes
-		if len(rbdVolumes) < int(i) {
-			return nil, status.Errorf(codes.Aborted, "invalid starting token %s", parseErr.Error())
+		if _, ok := err.(ErrImageNotFound); !ok {
+			return nil, status.Error(codes.Internal, err.Error())
 		}
-		startToken = int(i)
-	}
 
-	var entries []*csi.ListVolumesResponse_Entry
+		// If image is missing, then there was a key to point to it, it means we need to cleanup
+		// the keys and return success
+		volumeNameMutex.LockKey(rbdVol.RequestName)
+		defer func() {
+			if err := volumeNameMutex.UnlockKey(rbdVol.RequestName); err != nil {
+				klog.Warningf("failed to unlock mutex volume:%s %v", rbdVol.RequestName, err)
+			}
+		}()
 
-	keys := make([]string, 0)
-	for k := range rbdVolumes {
-		keys = append(keys, k)
+		if err := unreserveVol(rbdVol, req.GetSecrets()); err != nil {
```

Comment:
reserve and unreserveVol function name looks vague to me .

Reply:
may be some function names with `store` or `map`?

Reply:
In my refactor patch this is renamed as `ReserveName` and `UndoReservation` and `CheckReservation` for reservation of a name, undoing the same, and checking reservation of a name. I an NOT posting that patch yet as it is a refactor and I do not want to derail the review progress made here, but the patchis available [here](https://github.com/ShyamsundarR/ceph-csi/commit/c987e8bc7c8999e1072db1a902e64efb256abae1).

I am assuming this renaming is acceptable.